In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('/content/drive/MyDrive/Python Codes/Global company/companies_sorted.csv')
df.head()

Unnamed: 0                       name         domain  year founded  \
0     5872184                        ibm        ibm.com        1911.0   
1     4425416  tata consultancy services        tcs.com        1968.0   
2       21074                  accenture  accenture.com        1989.0   
3     2309813                    us army     goarmy.com        1800.0   
4     1558607                         ey         ey.com        1989.0   

                              industry size range  \
0  information technology and services     10001+   
1  information technology and services     10001+   
2  information technology and services     10001+   
3                             military     10001+   
4                           accounting     10001+   

                                 locality         country  \
0       new york, new york, united states   united states   
1              bombay, maharashtra, india           india   
2                 dublin, dublin, ireland         ireland   
3     alexandria, virginia, united states   united states   
4  london, greater london, united kingdom  united kingdom   

                                     linkedin url  current employee estimate  \
0                        linkedin.com/company/ibm                     274047   
1  linkedin.com/company/tata-consultancy-services                     190771   
2                  linkedin.com/company/accenture                     190689   
3                    linkedin.com/company/us-army                     162163   
4              linkedin.com/company/ernstandyoung                     158363   

   total employee estimate  
0                   716906  
1                   341369  
2                   455768  
3                   445958  
4                   428960

Finding out the shape of the dataset ---

In [5]:
df.shape

(7173426, 11)

So the dataset has 7 Million rows and 11 feature fields

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7173426 entries, 0 to 7173425
Data columns (total 11 columns):
 #   Column                     Dtype  
---  ------                     -----  
 0   Unnamed: 0                 int64  
 1   name                       object 
 2   domain                     object 
 3   year founded               float64
 4   industry                   object 
 5   size range                 object 
 6   locality                   object 
 7   country                    object 
 8   linkedin url               object 
 9   current employee estimate  int64  
 10  total employee estimate    int64  
dtypes: float64(1), int64(3), object(7)
memory usage: 602.0+ MB


In [7]:
df.isnull().any()

Unnamed: 0                   False
name                          True
domain                        True
year founded                  True
industry                      True
size range                   False
locality                      True
country                       True
linkedin url                 False
current employee estimate    False
total employee estimate      False
dtype: bool

Dropping unnamed and size range columns

In [8]:
df.drop(columns = ['Unnamed: 0', 'size range'], inplace = True)

# CLEANING

As in this project, we are mainly concerned about the names of the companies, so it has to be unique and not null value. Let's look into those -

In [9]:
df[df.name.isna()]

name       domain  year founded                             industry  \
192621   NaN       nan.nl           NaN                    apparel & fashion   
512291   NaN  nan-tic.com        2010.0  information technology and services   
2599907  NaN          NaN           NaN                    computer hardware   

                                     locality      country  \
192621   rotterdam, zuid-holland, netherlands  netherlands   
512291             sabadell, catalonia, spain        spain   
2599907                                   NaN          NaN   

                         linkedin url  current employee estimate  \
192621     linkedin.com/company/nan_2                         54   
512291   linkedin.com/company/nan-tic                         18   
2599907      linkedin.com/company/nan                          2   

         total employee estimate  
192621                        61  
512291                        29  
2599907                        2

In [10]:
df.loc[df.index == 512291, 'name'] = 'Nan-Tic'

In [11]:
df.dropna(axis = 0, subset = ['name'], inplace = True)

In [12]:
df.isna().sum()

name                               0
domain                       1650620
year founded                 3606978
industry                      290003
locality                     2508824
country                      2349206
linkedin url                       0
current employee estimate          0
total employee estimate            0
dtype: int64

Now we will see if there are duplicates in the name 

In [13]:
df.duplicated(subset = ['name']).all()

False

Another way to check duplicates is to do value counts of each record in the column.

In [14]:
df['name'].value_counts()

independent consultant            69
consultant                        58
private practice                  54
independent                       52
independent contractor            41
                                  ..
mvn future genius                  1
christiansen farms                 1
joliz coaching ltd                 1
boutique femme                     1
dynamic controls solutions inc     1
Name: name, Length: 7004635, dtype: int64

So, we can see that there are duplicates. let's check how these duplicates looks like and then we will remove it.

In [15]:
df[df['name'] == 'ibm']

name                         domain  year founded  \
0        ibm                        ibm.com        1911.0   
41433    ibm                            NaN           NaN   
1136735  ibm                            NaN        1912.0   
1904804  ibm                            NaN        1914.0   
2325002  ibm  mcgeemarketingconsultants.com           NaN   
4091427  ibm                   ezsource.com        2003.0   
6751962  ibm                            NaN           NaN   
6814019  ibm                 ibmconsult.com        2012.0   
7151813  ibm                            NaN           NaN   

                                    industry  \
0        information technology and services   
41433    information technology and services   
1136735  information technology and services   
1904804  information technology and services   
2325002                    computer software   
4091427                    computer software   
6751962                                  NaN   
6814019            marketing and advertising   
7151813                                  NaN   

                                  locality        country  \
0        new york, new york, united states  united states   
41433                                  NaN            NaN   
1136735                                NaN            NaN   
1904804    colombes, ile-de-france, france         france   
2325002                                NaN            NaN   
4091427     mevo modi`im, hamerkaz, israel         israel   
6751962                                NaN            NaN   
6814019                                NaN            NaN   
7151813                                NaN            NaN   

                                        linkedin url  \
0                           linkedin.com/company/ibm   
41433              linkedin.com/company/global-value   
1136735  linkedin.com/company/ibm-united-kingdom-ltd   
1904804              linkedin.com/company/ibm-france   
2325002                linkedin.com/company/mindflow   
4091427                linkedin.com/company/ezlegacy   
6751962                  linkedin.com/company/ibmdaf   
6814019          linkedin.com/company/ibmconsult-com   
7151813                linkedin.com/company/fiokware   

         current employee estimate  total employee estimate  
0                           274047                   716906  
41433                          280                      931  
1136735                          7                       50  
1904804                          3                        8  
2325002                          2                        6  
4091427                          1                       12  
6751962                          0                        1  
6814019                          0                        2  
7151813                          0                        1

In [16]:
df.drop_duplicates(subset=['name'], keep='first', inplace = True)

In [17]:
df[df['name'] == 'ibm']

name   domain  year founded                             industry  \
0  ibm  ibm.com        1911.0  information technology and services   

                            locality        country              linkedin url  \
0  new york, new york, united states  united states  linkedin.com/company/ibm   

   current employee estimate  total employee estimate  
0                     274047                   716906

Thus, we have a new dataframe without the duplicate names.

In [18]:
df.shape

(7004635, 9)

From rows 7173426, we have got 7004635 after dropping duplicates and keeping the first record of the duplicates.

Now let's look into the domains. We have seen that the domains have 1650620 null data. Let's cover those first.

In [19]:
df[df.domain.isna()]

name domain  \
32                                                     nhs    NaN   
57                        ayatama energi, trisco nusantara    NaN   
59                            american center of krasnodar    NaN   
61       promobroker agente de seguros y de fianzas s a...    NaN   
81                             verbum traducción y edición    NaN   
...                                                    ...    ...   
7173418                              koop media management    NaN   
7173419                                    la belle miette    NaN   
7173420                                inkredible printing    NaN   
7173423                         catholic bishop of chicago    NaN   
7173424                                medexo robotics ltd    NaN   

         year founded                             industry  \
32             1948.0               hospital & health care   
57             1999.0                               design   
59                NaN                 education management   
61                NaN                            insurance   
81                NaN         translation and localization   
...               ...                                  ...   
7173418           NaN                      broadcast media   
7173419           NaN                                  NaN   
7173420           NaN  information technology and services   
7173423           NaN               religious institutions   
7173424           NaN                             research   

                                   locality         country  \
32                                      NaN             NaN   
57                                      NaN             NaN   
59                                      NaN             NaN   
61                                      NaN             NaN   
81                                      NaN             NaN   
...                                     ...             ...   
7173418                                 NaN             NaN   
7173419                                 NaN             NaN   
7173420  artesia, new mexico, united states   united states   
7173423  inverness, illinois, united states   united states   
7173424      london, london, united kingdom  united kingdom   

                                              linkedin url  \
32                                linkedin.com/company/nhs   
57       linkedin.com/company/ayatama-energi-trisco-nus...   
59       linkedin.com/company/american-center-of-krasnodar   
61       linkedin.com/company/promobroker-agente-de-seg...   
81        linkedin.com/company/verbum-traducci-n-y-edici-n   
...                                                    ...   
7173418         linkedin.com/company/koop-media-management   
7173419               linkedin.com/company/la-belle-miette   
7173420           linkedin.com/company/inkredible-printing   
7173423    linkedin.com/company/catholic-bishop-of-chicago   
7173424           linkedin.com/company/medexo-robotics-ltd   

         current employee estimate  total employee estimate  
32                           67564                    92596  
57                           48806                    49370  
59                           47345                    47924  
61                           44926                    45655  
81                           39679                    39736  
...                            ...                      ...  
7173418                          0                        1  
7173419                          0                        1  
7173420                          0                        1  
7173423                          0                        1  
7173424                          0                        2  

[1569331 rows x 9 columns]

In [20]:
df.domain.replace(np.nan, 'Not Given', inplace = True)

In [21]:
df.domain.value_counts()

Not Given                       1569331
nordalps.com                         12
virtualcu.net                         8
play-cricket.com                      8
dunked.com                            8
                                 ...   
librandt.pl                           1
weststreetwellbeing.com.au            1
theadventurehotel.com                 1
sinergeticaitalia.it                  1
dynamiccontrolssolutions.com          1
Name: domain, Length: 5391226, dtype: int64

In [22]:
df[df['domain'] == 'nordalps.com']

name        domain  year founded  \
3306163  mazzoni enterprises  nordalps.com        2015.0   
3370492    red flame studios  nordalps.com        2015.0   
3405271     frost wolf media  nordalps.com        2015.0   
3505539    black fox records  nordalps.com        2015.0   
3869219    speed cat fitness  nordalps.com        2015.0   
3898088       blue fox radio  nordalps.com        2015.0   
3970499   green leaf academy  nordalps.com        2015.0   
4612172  cold harbor studios  nordalps.com        2015.0   
4864208               kubbix  nordalps.com        2015.0   
5015945            nord alps  nordalps.com        2015.0   
5357577     dark bat studios  nordalps.com        2015.0   
5363056  royal beats records  nordalps.com        2015.0   

                             industry                    locality country  \
3306163              media production      milan, lombardy, italy   italy   
3370492              media production      milan, lombardy, italy   italy   
3405271              media production      milan, lombardy, italy   italy   
3505539                         music  aosta, aosta valley, italy   italy   
3869219  health, wellness and fitness  aosta, aosta valley, italy   italy   
3898088                         music      milan, lombardy, italy   italy   
3970499                    e-learning      milan, lombardy, italy   italy   
4612172              media production      milan, lombardy, italy   italy   
4864208              media production      milan, lombardy, italy   italy   
5015945              media production      milan, lombardy, italy   italy   
5357577              media production      milan, lombardy, italy   italy   
5363056                         music      milan, lombardy, italy   italy   

                                     linkedin url  current employee estimate  \
3306163  linkedin.com/company/mazzoni-enterprises                          1   
3370492    linkedin.com/company/red-flame-studios                          1   
3405271     linkedin.com/company/frost-wolf-media                          1   
3505539    linkedin.com/company/black-fox-records                          1   
3869219    linkedin.com/company/speed-cat-fitness                          1   
3898088       linkedin.com/company/blue-fox-radio                          1   
3970499   linkedin.com/company/green-leaf-academy                          1   
4612172  linkedin.com/company/cold-harbor-studios                          1   
4864208               linkedin.com/company/kubbix                          1   
5015945            linkedin.com/company/nord-alps                          1   
5357577     linkedin.com/company/dark-bat-studios                          1   
5363056  linkedin.com/company/royal-beats-records                          1   

         total employee estimate  
3306163                        1  
3370492                        1  
3405271                        1  
3505539                        1  
3869219                        1  
3898088                        1  
3970499                        2  
4612172                        1  
4864208                        1  
5015945                        1  
5357577                        1  
5363056                        1

In [23]:
df['year founded'].fillna(0, inplace = True)

In [24]:
df.describe().transpose()

count        mean          std  min  25%  50%  \
year founded               7004635.0  998.638067  1001.001641  0.0  0.0  0.0   
current employee estimate  7004635.0   13.961687   358.286721  0.0  1.0  1.0   
total employee estimate    7004635.0   32.525435   882.574316  1.0  1.0  2.0   

                              75%       max  
year founded               2009.0    2103.0  
current employee estimate     4.0  274047.0  
total employee estimate       7.0  716906.0

As we can see that the maximum year is 2103, which is not possible hence there is something wrong in it. We will see the number of companies listed as per years.

In [25]:
year_data = df['year founded'].value_counts().rename_axis('years').reset_index(name='count')
year_data

years    count
0       0.0  3510216
1    2015.0   228478
2    2014.0   226894
3    2013.0   217440
4    2012.0   207421
..      ...      ...
229  2023.0        1
230  2029.0        1
231  1451.0        1
232  1800.0        1
233  1792.0        1

[234 rows x 2 columns]

In [26]:
df.drop(df.index[df['year founded'] > 2021], inplace=True)

In [27]:
df.shape

(7004624, 9)

In [28]:
df[df['year founded'] == 0].head()

name             domain  year founded  \
12                   wells fargo     wellsfargo.com           0.0   
18                       us navy           navy.com           0.0   
24           education nationale  education.gouv.fr           0.0   
29  united states postal service           usps.com           0.0   
31                      ericsson       ericsson.com           0.0   

                               industry  \
12                   financial services   
18                             military   
24                        public policy   
29            government administration   
31  information technology and services   

                                           locality        country  \
12         san francisco, california, united states  united states   
18  washington, district of columbia, united states  united states   
24                     paris, île-de-france, france         france   
29  washington, district of columbia, united states  united states   
31                         kista, stockholm, sweden         sweden   

                                linkedin url  current employee estimate  \
12           linkedin.com/company/wellsfargo                     109532   
18              linkedin.com/company/us-navy                      94458   
24  linkedin.com/company/education-nationale                      84218   
29                 linkedin.com/company/usps                      74357   
31             linkedin.com/company/ericsson                      67692   

    total employee estimate  
12                   264101  
18                   325740  
24                   101083  
29                   104159  
31                   178578

In [29]:
current_employee = df['current employee estimate'].value_counts().rename_axis('current employee').reset_index(name='count')
current_employee

current employee    count
0                    1  2387375
1                    0  1414859
2                    2   944639
3                    3   483871
4                    4   300416
...                ...      ...
5348              4590        1
5349              4584        1
5350              4580        1
5351              4575        1
5352              2784        1

[5353 rows x 2 columns]

In [30]:
df.query("0 < `current employee estimate` < 100")['current employee estimate'].value_counts().rename_axis('current employee').reset_index(name='count')

current employee    count
0                  1  2387375
1                  2   944639
2                  3   483871
3                  4   300416
4                  5   206873
..               ...      ...
94                99     1104
95                95     1100
96                98     1075
97                97     1057
98                94     1054

[99 rows x 2 columns]

In [31]:
df[df['current employee estimate'] == 1].head()

name                     domain  \
3265732             ict insight pty ltd          ictinsight.com.au   
3265733     linden crest surgery center         lindensurgical.com   
3265734                 spreadsheetgear        spreadsheetgear.com   
3265735            office&cowork centre                  Not Given   
3265736  the sign and poster studio ltd  signandposterstudio.co.uk   

         year founded                             industry  \
3265732           0.0  information technology and services   
3265733           0.0               hospital & health care   
3265734        2003.0                    computer software   
3265735           0.0                investment management   
3265736           0.0                             printing   

                                         locality         country  \
3265732                                       NaN             NaN   
3265733  beverly hills, california, united states   united states   
3265734             lenexa, kansas, united states   united states   
3265735                                       NaN             NaN   
3265736    lancashire, lancashire, united kingdom  united kingdom   

                                              linkedin url  \
3265732           linkedin.com/company/ict-insight-pty-ltd   
3265733   linkedin.com/company/linden-crest-surgery-center   
3265734               linkedin.com/company/spreadsheetgear   
3265735          linkedin.com/company/office-cowork-centre   
3265736  linkedin.com/company/the-sign-and-poster-studi...   

         current employee estimate  total employee estimate  
3265732                          1                        1  
3265733                          1                        2  
3265734                          1                        1  
3265735                          1                        1  
3265736                          1                        1

In [32]:
df.drop(df.index[df['current employee estimate'] < 100], inplace=True)

In [33]:
df.shape

(108294, 9)

In [34]:
df.isna().sum()

name                             0
domain                           0
year founded                     0
industry                         0
locality                     23462
country                      21494
linkedin url                     0
current employee estimate        0
total employee estimate          0
dtype: int64

In [35]:
df.describe().transpose()

count         mean          std    min    25%  \
year founded               108294.0  1362.611484   910.670408    0.0    0.0   
current employee estimate  108294.0   621.728184  2814.511915  100.0  135.0   
total employee estimate    108294.0  1425.635742  6953.429765  100.0  283.0   

                              50%     75%       max  
year founded               1960.0  1994.0    2019.0  
current employee estimate   208.0   425.0  274047.0  
total employee estimate     467.0   966.0  716906.0

In [36]:
df[df['country'].isna() & df['locality'].isna()]

name  \
32                                                    nhs   
54                                                    sap   
57                       ayatama energi, trisco nusantara   
59                           american center of krasnodar   
61      promobroker agente de seguros y de fianzas s a...   
...                                                   ...   
110861                                               orda   
110867                                          coteccons   
110868                                  consórcio realiza   
110876        上海猎益信息技术有限公司 shanghai recruit easy co., ltd   
110879                                      huyck.wangner   

                         domain  year founded  \
32                    Not Given        1948.0   
54                      sap.com        1972.0   
57                    Not Given        1999.0   
59                    Not Given           0.0   
61                    Not Given           0.0   
...                         ...           ...   
110861                Not Given           0.0   
110867             coteccons.vn        2004.0   
110868  consorciorealiza.com.br        1993.0   
110876               hr2002.com        2002.0   
110879              wangner.com           0.0   

                                    industry locality country  \
32                    hospital & health care      NaN     NaN   
54                         computer software      NaN     NaN   
57                                    design      NaN     NaN   
59                      education management      NaN     NaN   
61                                 insurance      NaN     NaN   
...                                      ...      ...     ...   
110861  recreational facilities and services      NaN     NaN   
110867                          construction      NaN     NaN   
110868                    financial services      NaN     NaN   
110876               staffing and recruiting      NaN     NaN   
110879               paper & forest products      NaN     NaN   

                                             linkedin url  \
32                               linkedin.com/company/nhs   
54                               linkedin.com/company/sap   
57      linkedin.com/company/ayatama-energi-trisco-nus...   
59      linkedin.com/company/american-center-of-krasnodar   
61      linkedin.com/company/promobroker-agente-de-seg...   
...                                                   ...   
110861                          linkedin.com/company/orda   
110867                     linkedin.com/company/coteccons   
110868             linkedin.com/company/consorcio-realiza   
110876                  linkedin.com/company/上海猎益信息技术有限公司   
110879                 linkedin.com/company/huyck.wangner   

        current employee estimate  total employee estimate  
32                          67564                    92596  
54                          51441                   102773  
57                          48806                    49370  
59                          47345                    47924  
61                          44926                    45655  
...                           ...                      ...  
110861                        100                      149  
110867                        100                      150  
110868                        100                      169  
110876                        100                      104  
110879                        100                      202  

[21494 rows x 9 columns]

In [37]:
df.dropna(axis = 0, subset = ['country'], inplace = True)

In [38]:
df.shape

(86800, 9)

In [39]:
df.isna().sum()

name                            0
domain                          0
year founded                    0
industry                        0
locality                     1968
country                         0
linkedin url                    0
current employee estimate       0
total employee estimate         0
dtype: int64

In [41]:
df.locality.replace(np.nan, 'Not Given', inplace = True)

In [42]:
df.isna().sum()

name                         0
domain                       0
year founded                 0
industry                     0
locality                     0
country                      0
linkedin url                 0
current employee estimate    0
total employee estimate      0
dtype: int64

In [43]:
df.shape

(86800, 9)

In [46]:
df.to_csv('/content/drive/MyDrive/Python Codes/Global company/filtered_file.csv', index = False)